debian/patches/0069-add-tonemap-videotoolbox.patch

```diff
@@ -0,0 +1,1867 @@
++}
++
++float gamma(float s, float peak, float target_peak) {
++    float p = s > 0.05f ? s / peak : 0.05f / peak;
++    float v = powr(p, 1.0f / tone_param);
++    return s > 0.05f ? v : (s * v / 0.05f);
++}
++
++float clip(float s, float peak, float target_peak) {
++    return clamp(s * tone_param, 0.0f, 1.0f);
++}
++
++float reinhard(float s, float peak, float target_peak) {
++    return s / (s + tone_param) * (peak + tone_param) / peak;
++}
++
++float hable(float s, float peak, float target_peak) {
++    return hable_f(s) / hable_f(peak);
++}
++
++float mobius(float s, float peak, float target_peak) {
++    float j = tone_param;
++    float a, b;
++
++    if (s <= j)
++        return s;
++
++    a = -j * j * (peak - 1.0f) / (j * j - 2.0f * j + peak);
++    b = (j * j - 2.0f * j * peak + peak) / fmax(peak - 1.0f, FLOAT_EPS);
++
++    return (b * b + 2.0f * b * j + j * j) / (b - a) * (s + a) / (s + b);
++}
++
++float bt2390(float s, float peak_inv_pq, float target_peak_inv_pq) {
++    float peak_pq = peak_inv_pq;
++    float scale = peak_pq > 0.0f ? (1.0f / peak_pq) : 1.0f;
++
++    float s_pq = inverse_eotf_st2084(s) * scale;
++    float max_lum = target_peak_inv_pq * scale;
++
++    float ks = 1.5f * max_lum - 0.5f;
++    float tb = (s_pq - ks) / (1.0f - ks);
++    float tb2 = tb * tb;
++    float tb3 = tb2 * tb;
++    float pb = (2.0f * tb3 - 3.0f * tb2 + 1.0f) * ks +
++               (tb3 - 2.0f * tb2 + tb) * (1.0f - ks) +
++               (-2.0f * tb3 + 3.0f * tb2) * max_lum;
++    float sig = mix(pb, s_pq, s_pq < ks);
++
++    return eotf_st2084(sig * peak_pq);
++}
++
++float tonemap(float s, float peak, float target_peak) {
++    if (tonemap_func_type == 1) {
++        return direct(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 2) {
++        return linear(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 3) {
++        return gamma(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 4) {
++        return clip(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 5) {
++        return reinhard(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 6) {
++        return hable(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 7) {
++        return mobius(s, peak, target_peak);
++    }
++    if (tonemap_func_type == 8) {
++        return bt2390(s, peak, target_peak);
++    }
++    return direct(s, peak, target_peak);
++}
++
++float get_dithered_y(float y, float d) {
++    return floor(y * dither_quantization + d + 0.5f / dither_size2) * 1.0f / dither_quantization;
++}
++
++void map_four_pixels_rgb(thread float4 *r4, thread float4 *g4, thread float4 *b4, float peak) {
++#define MAP_FOUR_PIXELS(sig, peak, target_peak) \
++{ \
++    sig.x = tonemap(sig.x, peak, target_peak); \
++    sig.y = tonemap(sig.y, peak, target_peak); \
++    sig.z = tonemap(sig.z, peak, target_peak); \
++    sig.w = tonemap(sig.w, peak, target_peak); \
++}
++    if (is_tone_mode_rgb) {
```

Comment:
Will this runtime if/else check cause branch divergence on Metal?
I know Cuda doesn't like this very much, so I have to create kernels for every path.

Reply:
This isn't a runtime if-else check after the Metal magic. The `is_tone_mode_rgb` is referred to as *function constants* and is a value to be set at runtime but stays constant during each function calls. When we are creating the Metal function with function constants, the Metal backend compiler will compile the specialized function at runtime, eliminating all constant branches with `if(false)`. When this is uploaded to the GPU, only the branch evaluated to true will remain.
https://github.com/gnattu/jellyfin-ffmpeg/blob/9b2975da5170275d5715c4be7e3367432a4a1729/debian/patches/0069-add-tonemap-videotoolbox.patch#L1433

Reply:
Then it seems that Cuda doesn't do link-time optimization. cuLink (constants ptx + tonemap ptx), because ptx compilation occurs when ffmpeg is compiled.

So, when I use `__constant__ cond` for if(cond)...else in Cuda, there is still a performance penalty compared to no branching.

Reply:
On Apple Silicon Macs the performance is currently limited by the hardware decoder/encoders or whatever throttling Apple is putting on it. Even removing the tone-map filter altogether the fps does not increase, so our kernel is doing its best now.